core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala

```diff
@@ -0,0 +1,204 @@
+
+    conf.setMaster(dispatcherArgs.masterUrl)
+    conf.setAppName("Mesos Cluster Dispatcher")
+
+    dispatcherArgs.zookeeperUrl.foreach {
+      z => {
+        conf.set("spark.deploy.recoveryMode", "ZOOKEEPER")
+        conf.set("spark.deploy.zookeeper.url", z)
+      }
+    }
+
+    val dispatcher = new MesosClusterDispatcher(
+      dispatcherArgs,
+      conf)
+
+    dispatcher.start()
+
+    val shutdownHook = new Thread() {
+      override def run() {
+        logInfo("Shutdown hook is shutting down dispatcher")
+        dispatcher.stop()
+        dispatcher.awaitShutdown()
+      }
+    }
+
+    Runtime.getRuntime.addShutdownHook(shutdownHook)
+
+    dispatcher.awaitShutdown()
+  }
+
+  class ClusterDispatcherArguments(args: Array[String], conf: SparkConf) {
+    var host = Utils.localHostName()
+    var port = 7077
+    var webUiPort = 8081
+    var masterUrl: String = _
+    var zookeeperUrl: Option[String] = None
+    var propertiesFile: String = _
+
+    parse(args.toList)
+
+    propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
+
+    def parse(args: List[String]): Unit = args match {
+      case ("--host" | "-h") :: value :: tail =>
+        Utils.checkHost(value, "Please use hostname " + value)
+        host = value
+        parse(tail)
+
+      case ("--port" | "-p") :: IntParam(value) :: tail =>
+        port = value
+        parse(tail)
+
+      case ("--webui-port" | "-p") :: IntParam(value) :: tail =>
+        webUiPort = value
+        parse(tail)
+
+      case ("--zk" | "-z") :: value :: tail =>
+        zookeeperUrl = Some(value)
+        parse(tail)
+
+      case ("--master" | "-m") :: value :: tail =>
+        if (!value.startsWith("mesos://")) {
+          System.err.println("Cluster dispatcher only supports mesos (uri begins with mesos://)")
+          System.exit(1)
+        }
+        masterUrl = value.stripPrefix("mesos://")
+        parse(tail)
+
+      case ("--properties-file") :: value :: tail =>
+        propertiesFile = value
+        parse(tail)
+
+      case ("--help") :: tail =>
+        printUsageAndExit(0)
+
+      case Nil => {
+        if (masterUrl == null) {
+          System.err.println("--master is required")
+          System.exit(1)
+        }
+      }
+
+      case _ =>
+        printUsageAndExit(1)
+    }
+
+    /**
+     * Print usage and exit JVM with the given exit code.
+     */
+    def printUsageAndExit(exitCode: Int) {
+      System.err.println(
+        "Usage: MesosClusterDispatcher [options]\n" +
+          "\n" +
+          "Options:\n" +
+          "  -h HOST, --host HOST   Hostname to listen on\n" +
+          "  -p PORT, --port PORT   Port to listen on (default: 7077)\n" +
+          "  --webui-port WEBUI_PORT   WebUI Port to listen on (default: 8081)\n" +
+          "  -m --master MASTER      URI for connecting to Mesos master\n" +
+          "  -z --zk ZOOKEEPER       URLs for connecting to Zookeeper for persistence\n" +
```

Comment:
is this comma-delimited? If so, we should say so.
Also, looks like the spacing here isn't really aligned.